core/src/main/java/hudson/cli/CLIAction.java

```diff
@@ -91,46 +107,159 @@ public Object getTarget() {
+                throw new PlainCliEndpointResponse();
+            } else if (jenkins.CLI.get().isEnabled()) {
+                throw new RemotingCliEndpointResponse();
+            } else {
+                throw HttpResponses.forbidden();
+            }
         } else {
             return this;
         }
     }
 
     /**
-     * Serves CLI-over-HTTP response.
+     * Serves {@link PlainCLIProtocol} response.
      */
-    private class CliEndpointResponse extends HttpResponseException {
-        @Override
-        public void generateResponse(StaplerRequest req, StaplerResponse rsp, Object node) throws IOException, ServletException {
-            try {
-                // do not require any permission to establish a CLI connection
-                // the actual authentication for the connecting Channel is done by CLICommand
+    private class PlainCliEndpointResponse extends FullDuplexHttpService.Response {
 
-                UUID uuid = UUID.fromString(req.getHeader("Session"));
-                rsp.setHeader("Hudson-Duplex",""); // set the header so that the client would know
+        PlainCliEndpointResponse() {
+            super(duplexServices);
+        }
 
-                FullDuplexHttpChannel server;
-                if(req.getHeader("Side").equals("download")) {
-                    duplexChannels.put(uuid,server=new FullDuplexHttpChannel(uuid, !Jenkins.getActiveInstance().hasPermission(Jenkins.ADMINISTER)) {
+        @Override
+        protected FullDuplexHttpService createService(StaplerRequest req, UUID uuid) throws IOException {
+            return new FullDuplexHttpService(uuid) {
+                @Override
+                protected void run(InputStream upload, OutputStream download) throws IOException, InterruptedException {
+                    final AtomicReference<Thread> runningThread = new AtomicReference<>();
+                    class ServerSideImpl extends PlainCLIProtocol.ServerSide {
+                        boolean ready;
+                        List<String> args = new ArrayList<>();
+                        Locale locale = Locale.getDefault();
+                        Charset encoding = Charset.defaultCharset();
+                        final PipedInputStream stdin = new PipedInputStream();
+                        final PipedOutputStream stdinMatch = new PipedOutputStream();
+                        ServerSideImpl(InputStream is, OutputStream os) throws IOException {
+                            super(is, os);
+                            stdinMatch.connect(stdin);
+                        }
+                        @Override
+                        protected void onArg(String text) {
+                            args.add(text);
+                        }
+                        @Override
+                        protected void onLocale(String text) {
+                            for (Locale _locale : Locale.getAvailableLocales()) {
+                                if (_locale.toString().equals(text)) {
```

Comment:
Yes according to Oracle's documentation: https://docs.oracle.com/javase/7/docs/api/java/util/Locale.html#toString() . I am fine with keeping it as is